File: cc/ping_list.go
// go run ping_report.go [list_file] [count]
  package main

  import (
      "bufio"
      "fmt"
      "math"
      "os"
      "os/exec"
      "regexp"
      "sort"
      "strconv"
      "strings"
      "time"
  )

  type sample struct {
      target string
      ok     bool
      lat    float64
  }

  type summary struct {
      Target string
      Sent   int
      Recv   int
      Loss   float64
      Min    float64
      Avg    float64
      Max    float64
      Mdev   float64
  }

  var timeRE = regexp.MustCompile(`time[=<]\s*([\d.]+)\s*ms`)

  func pingOnce(target string, timeout int) (bool, float64) {
      cmd := exec.Command("ping", "-c", "1", "-W", strconv.Itoa(timeout), target)
      out, err := cmd.Output()
      if err != nil {
          return false, 0
      }
      m := timeRE.FindStringSubmatch(string(out))
      if len(m) < 2 {
          return false, 0
      }
      lat, err := strconv.ParseFloat(m[1], 64)
      if err != nil {
          return false, 0
      }
      return true, lat
  }

  func probe(target string, count, timeout int) []sample {
      samples := make([]sample, 0, count)
      for i := 0; i < count; i++ {
          ok, lat := pingOnce(target, timeout)
          samples = append(samples, sample{target, ok, lat})
      }
      return samples
  }

  func summarize(samples []sample) summary {
      var s summary
      s.Target = samples[0].target
      s.Sent = len(samples)
      var vals []float64
      for _, x := range samples {
          if x.ok {
              s.Recv++
              vals = append(vals, x.lat)
          }
      }
      s.Loss = float64(s.Sent-s.Recv) / float64(s.Sent) * 100
      if len(vals) > 0 {
          s.Min, s.Max = vals[0], vals[0]
          var sum float64
          for _, v := range vals {
              sum += v
              if v < s.Min {
                  s.Min = v
              }
              if v > s.Max {
                  s.Max = v
              }
          }
          s.Avg = sum / float64(len(vals))
          if len(vals) > 1 {
              var sq float64
              for _, v := range vals {
                  sq += (v - s.Avg) * (v - s.Avg)
              }
              s.Mdev = math.Sqrt(sq / float64(len(vals)-1))
          }
      }
      return s
  }

  func loadTargets(path string) ([]string, error) {
      f, err := os.Open(path)
      if err != nil {
          return nil, err
      }
      defer f.Close()
      var out []string
      sc := bufio.NewScanner(f)
      for sc.Scan() {
          t := strings.TrimSpace(sc.Text())
          if t == "" || strings.HasPrefix(t, "#") {
              continue
          }
          out = append(out, t)
      }
      return out, sc.Err()
  }

  func fmtOrDash(v float64) string {
      if v == 0 {
          return "-"
      }
      return strconv.FormatFloat(v, 'f', 2, 64)
  }

  func main() {
      listFile := "test_list"
      count := 4
      timeout := 2
      if len(os.Args) > 1 {
          listFile = os.Args[1]
      }
      if len(os.Args) > 2 {
          if n, err := strconv.Atoi(os.Args[2]); err == nil {
              count = n
          }
      }

      targets, err := loadTargets(listFile)
      if err != nil {
          fmt.Fprintln(os.Stderr, "打开文件失败:", err)
          os.Exit(1)
      }
      if len(targets) == 0 {
          fmt.Fprintln(os.Stderr, "无有效目标")
          os.Exit(0)
      }

      fmt.Printf("目标数: %d  每目标 ping次数: %d  超时: %ds\n", len(targets), count, timeout)
      fmt.Println("探测中...")

      summaries := make([]summary, 0, len(targets))
      for _, t := range targets {
          s := summarize(probe(t, count, timeout))
          summaries = append(summaries, s)
      }

      fmt.Printf("\n=== Ping 报告 === %s  每个目标 ping %d 次\n\n",
          time.Now().Format("2006-01-02 15:04:05"), count)
      fmt.Printf("%-24s %4s %4s %7s %9s %9s %9s %9s\n",
          "目标", "发送", "收到", "丢包%", "最小ms", "平均ms", "最大ms", "mdevms")
      fmt.Println(strings.Repeat("-", 80))

      byLat := append([]summary{}, summaries...)
      sort.SliceStable(byLat, func(i, j int) bool {
          if byLat[i].Loss != byLat[j].Loss {
              return byLat[i].Loss < byLat[j].Loss
          }
          return byLat[i].Avg < byLat[j].Avg
      })
      for _, s := range byLat {
          fmt.Printf("%-24s %4d %4d %6.1f%% %9s %9s %9s %9s\n",
              s.Target, s.Sent, s.Recv, s.Loss, fmtOrDash(s.Min), fmtOrDash(s.Avg),
              fmtOrDash(s.Max), fmtOrDash(s.Mdev))
      }

      fmt.Println("\n--- 延迟排行 (按平均延迟升序, 排除全丢包) ---")
      alive := append([]summary{}, summaries...)
      sort.SliceStable(alive, func(i, j int) bool { return alive[i].Avg < alive[j].Avg })
      i := 0
      for _, s := range alive {
          if s.Recv == 0 {
              continue
          }
          i++
          fmt.Printf("  %2d. %-24s 平均 %.2f ms  丢包 %.0f%%\n", i, s.Target, s.Avg, s.Loss)
      }

      fmt.Println("\n--- 丢包排行 (按丢包率降序) ---")
      byLoss := append([]summary{}, summaries...)
      sort.SliceStable(byLoss, func(i, j int) bool { return byLoss[i].Loss > byLoss[j].Loss })
      for k, s := range byLoss {
          flag := ""
          if s.Recv == 0 {
              flag = " ★全丢"
          }
          fmt.Printf("  %2d. %-24s 丢包 %.0f%% (%d/%d)%s\n",
              k+1, s.Target, s.Loss, s.Sent-s.Recv, s.Sent, flag)
      }
  }

File: cc/ping_list.py
#!/usr/bin/env python3
"""
读取 test_list，对每个 IP/域名执行 ping，统计延迟和丢包率，排行后输出报告。
"""

import subprocess
import re
import sys
import platform
from statistics import mean, stdev
from datetime import datetime

LIST_FILE = "test_list"
PING_COUNT = 4
TIMEOUT = 2  # 单次 ping 超时（秒）


def ping_once(target):
    """调用系统 ping 一次，返回 (success: bool, latency_ms: float|None)。"""
    system = platform.system().lower()
    if system == "windows":
        cmd = ["ping", "-n", "1", "-w", str(TIMEOUT * 1000), target]
    else:
        cmd = ["ping", "-c", "1", "-W", str(TIMEOUT), target]
    try:
        out = subprocess.run(cmd, capture_output=True, text=True, timeout=TIMEOUT + 2).stdout
    except (subprocess.TimeoutExpired, Exception):
        return False, None
    m = re.search(r"time[=<]\s*([\d.]+)\s*ms", out)
    if m:
        return True, float(m.group(1))
    return False, None


def probe(target, count):
    """对单个目标 ping count 次。返回 dict 含延迟统计与丢包率。"""
    samples = []
    success = 0
    for _ in range(count):
        ok, lat = ping_once(target)
        if ok:
            samples.append(lat)
            success += 1
    loss_pct = (count - success) / count * 100
    return {
        "target": target,
        "sent": count,
        "recv": success,
        "loss_pct": loss_pct,
        "min": min(samples) if samples else None,
        "avg": mean(samples) if samples else None,
        "max": max(samples) if samples else None,
        "mdev": stdev(samples) if len(samples) > 1 else 0.0,
 "samples": samples,
    }


def fmt(v, suf=""):
    return f"{v:.2f}{suf}" if v is not None else "—"


def load_targets(path):
    targets = []
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            t = line.strip()
            if not t or t.startswith("#"):
                continue
            targets.append(t)
    return targets


def print_report(results, count):
    print(f"\n=== Ping报告 === 时间: {datetime.now():%Y-%m-%d %H:%M:%S}  每个目标 ping {count} 次\n")
    header = f"{'目标':<24}{'发送':>4}{'收到':>4}{'丢包%':>8}{'最小ms':>9}{'平均ms':>9}{'最大ms':>9}{'mdevms':>9}"
    print(header)
    print("-" * len(header.encode("gbk", "ignore") or header.encode()))
    print("-" * 80)
    for r in sorted(results, key=lambda x: (x["loss_pct"], x["avg"] or 1e9)):
        print(
            f"{r['target']:<24}{r['sent']:>4}{r['recv']:>4}{r['loss_pct']:>8.1f}"
            f"{fmt(r['min']):>9}{fmt(r['avg']):>9}{fmt(r['max']):>9}{fmt(r['mdev']):>9}"
        )

    print("\n---延迟排行 (按平均延迟升序，排除全丢包) ---")
    alive = [r for r in results if r["recv"] > 0]
    for i, r in enumerate(sorted(alive, key=lambda x: x["avg"]), 1):
        print(f"  {i:>2}. {r['target']:<24} 平均 {r['avg']:.2f} ms  丢包 {r['loss_pct']:.0f}%")

    print("\n--- 丢包排行 (按丢包率降序) ---")
    for i, r in enumerate(sorted(results, key=lambda x: -x["loss_pct"]), 1):
        flag = " ★全丢" if r["recv"] == 0 else ""
        print(f"  {i:>2}. {r['target']:<24} 丢包 {r['loss_pct']:.0f}% ({r['sent']-r['recv']}/{r['sent']}){flag}")


def main():
    count = PING_COUNT
    list_file = LIST_FILE
    if len(sys.argv) >= 2:
        list_file = sys.argv[1]
    if len(sys.argv) >= 3:
        count = int(sys.argv[2])

    targets = load_targets(list_file)
    if not targets:
        print(f"{list_file} 中没有可用条目。")
        return    print(f"开始探测 {len(targets)} 个目标，每个 {count} 次 ...")
    results = [probe(t, count) for t in targets]
    print_report(results, count)


if __name__ == "__main__":
    main()
